packages/services/src/contents/index.ts

```diff
@@ -1080,6 +1108,29 @@ export class Drive implements Contents.IDrive {
     this._apiEndpoint = options.apiEndpoint ?? SERVICE_DRIVE_URL;
     this.serverSettings =
       options.serverSettings ?? ServerConnection.makeSettings();
+    this.contentProviderRegistry.drive = this;
+  }
+
+  private static _contentProviderRegistry:
+    | IContentProviderRegistry
+    | undefined = undefined;
+
+  static getContentProviderRegistry(): IContentProviderRegistry {
+    if (Drive._contentProviderRegistry === undefined) {
+      const contentProviderRegistry = new ContentProviderRegistry();
+      const restContentProvider = new RestContentProvider();
+      contentProviderRegistry.register(restContentProvider);
+      Drive._contentProviderRegistry = contentProviderRegistry;
+    }
+    return Drive._contentProviderRegistry;
+  }
+
+  get contentProviderRegistry(): IContentProviderRegistry {
+    const contentProviderRegistry = Drive.getContentProviderRegistry();
+    if (contentProviderRegistry.drive === undefined) {
+      contentProviderRegistry.drive = this;
+    }
+    return contentProviderRegistry;
```

Comment:
Second thoughts: _should_ the registry be per-drive rather than static? Why?

I'm thinking that if we have a specialized network drive, for the use case of RTC it would not work right?

So instead of `Drive.getContentProviderRegistry()` the collaboration plugin would call `contentsManager.defaultDrive.getContentProviderRegistry()`?

We would need to expose the `defaultDrive` (currently a private `_defaultDrive`) but this is much smaller cost.

Reply:
But isn't it too late if we register a content provider once the drive is instantiated? At that time, it could already have been used to access a file system, right?
Also, how would the plugin require the contents manager?

Reply:
> But isn't it too late if we register a content provider once the drive is instantiated

Good point, I think we should work to ensure the RTC plugin gets loaded early. That said if you are calling `Drive.getContentProviderRegistry()` in RTC extension, it would have exactly the same timing so I don't think this is an argument against this approach.

> Also, how would the plugin require the contents manager?

`app.serviceManager.contents` is the `ContentsManager` class used by the application. Does that work for you?

Reply:
> Good point, I think we should work to ensure the RTC plugin gets loaded early.

That's why there should be relationships between dependencies, and a drive should require a content provider registry, ensuring the former is created after the latter. My understanding was that JupyterLab only "starts" after all plugins have been started, meaning that by the time the RTC plugin has started, it has registered its content provider in the default `Drive`, which is instantiated later. But maybe that's a false expectation?

> `app.serviceManager.contents` is the `ContentsManager` class used by the application.

Yes I could probably use that to access the drive, if we decide to register content providers on the drive instance rather than on the drive class. The downside is that they would not be registered on other instances, but I'm having a hard time understanding what this means.

Reply:
>  The downside is that they would not be registered on other instances, but I'm having a hard time understanding what this means.

I think this is an upside in the sense that if another extensions was using adding a drive (e.g. a network drive, similar to what RTC did, but with files actually outside of jupyter-server control) then installing collaboration would not would not break it. If we hijack all drives by registering collaboration on class level then existing drives which provide content from outside of jupyter-server would break.

Reply:
Makes sense, since drives and content providers are orthogonal. I'll make that change.